src/llm/model.py

```diff
@@ -0,0 +1,285 @@
+        self.dim_k = dim_k
+
+    def forward(
+        self,
+        query: Float32[Tensor, "batch seq_len model_dim"],
+        key: Float32[Tensor, "batch seq_len model_dim"],
+        value: Float32[Tensor, "batch seq_len model_dim"],
+        is_causal: bool = False,
+    ) -> Float32[Tensor, "batch seq_len dim_v"]:
+        # project key and query
+        key = self.W_key(key)
+        query = self.W_query(query)
+
+        similarity = torch.matmul(query, torch.permute(key, dims=(0, 2, 1)))
+        scaled_similarity = torch.divide(similarity, math.sqrt(self.dim_k))
+
+        if is_causal:
+            scaled_similarity = scaled_similarity.masked_fill(
+                mask=torch.triu(torch.ones_like(scaled_similarity).bool(), diagonal=1),
+                value=-torch.inf,
+            )
+        attention_scores = torch.softmax(
+            scaled_similarity, dim=-1
+        )  # Careful with the `dim`
+
+        # project the value
+        value = self.W_value(value)
+        outputs = torch.matmul(attention_scores, value)
+
+        return outputs
+
+
+class MultiHeadAttention(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_heads: int,
+    ):
+        super().__init__()
+        dim_k = dim_v = model_dim // num_heads
+        self.projection_heads = list()
+        for _ in range(num_heads):
+            self.projection_heads.append(
+                ScaledDotProductAttention(model_dim=model_dim, dim_k=dim_k, dim_v=dim_v)
+            )
+        self.W_out = nn.Linear(model_dim, model_dim)
+
+    def forward(
+        self,
+        query: Float32[Tensor, "batch seq_len model_dim"],
+        key: Float32[Tensor, "batch seq_len model_dim"],
+        value: Float32[Tensor, "batch seq_len model_dim"],
+        is_causal: bool=False,
+    ) -> Float32[Tensor, "batch seq_len model_dim"]:
+        outputs = list()
+        for head in self.projection_heads:
+            outputs.append(head(query, key, value, is_causal=is_causal))
+
+        outputs = torch.concat(outputs, dim=-1)
+        outputs = self.W_out(outputs)
+        return outputs
+
+
+class EncoderLayer(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_heads: int,
+        expansion_dim: int,
+    ):
+        super().__init__()
+        self.multi_head_attention = MultiHeadAttention(
+            model_dim=model_dim, num_heads=num_heads
+        )
+        self.feed_forward = FeedForward(
+            model_dim=model_dim, expansion_dim=expansion_dim
+        )
+        self.layer_norm1 = nn.LayerNorm(normalized_shape=model_dim)
+        self.layer_norm2 = nn.LayerNorm(normalized_shape=model_dim)
+
+    def forward(
+        self, inputs: Float32[Tensor, "batch enc_seq_len model_dim"]
+    ) -> Float32[Tensor, "batch enc_seq_len model_dim"]:
+        residual = inputs
+
+        x = self.multi_head_attention(query=inputs, key=inputs, value=inputs)
+        x = self.layer_norm1(x) + residual
```

Comment:
The original formulation is: `LayerNorm(input + sublayer(input)`

here it should be: `x = self.layer_norm1(residual + x)`

Reply:
Super catch!